File: services/user-service/src/infrastructure/messaging/KafkaEventPublisher.ts
import { Kafka } from "kafkajs";
import type { IEventPublisher } from "../../application/ports/IEventPublisher.js";

export class KafkaEventPublisher implements IEventPublisher {
    private producer: any;

    constructor() {
        const kafka = new Kafka({
            clientId: "user-service",
            brokers: (process.env.KAFKA_BROKER || "localhost:9092").split(","),
        });

        this.producer = kafka.producer();
    }

    async connect(): Promise<void> {
        await this.producer.connect();
    }

    async disconnect(): Promise<void> {
        await this.producer.disconnect();
    }

    async publish(eventName: string, payload: Record<string, unknown>): Promise<void> {
        try {
            await this.producer.send({
                topic: eventName,
                messages: [
                    {
                        key: (payload.userId as string) || (payload.entityId as string) || "unknown",
                        value: JSON.stringify({
                            ...payload,
                            eventType: eventName,
                            timestamp: new Date().toISOString(),
                        }),
                    },
                ],
            });
        } catch (error) {
            console.error("Error publishing event:", error);
            throw error;
        }
    }
}

File: services/ai-chat-service/src/infrastructure/messaging/kafka_event_publisher.py
from kafka import KafkaProducer
from typing import Dict, Any
import json
import os
from src.application.ports.ievent_publisher import IEventPublisher


class KafkaEventPublisher(IEventPublisher):
    def __init__(self):
        kafka_broker = os.getenv("KAFKA_BROKER", "localhost:9092")
        self.producer = KafkaProducer(
            bootstrap_servers=[kafka_broker],
            value_serializer=lambda v: json.dumps(v).encode("utf-8"),
            key_serializer=lambda k: k.encode("utf-8") if k else None,
        )

    async def connect(self) -> None:
        # KafkaProducer se conecta automáticamente
        pass

    async def disconnect(self) -> None:
        self.producer.close()

    async def publish(self, event_name: str, payload: Dict[str, Any]) -> None:
        try:
            key = payload.get("promptId") or payload.get("userId") or "unknown"
            value = {
                **payload,
                "eventType": event_name,
                "timestamp": __import__("datetime").datetime.utcnow().isoformat(),
            }
            self.producer.send(event_name, key=key, value=value)
            self.producer.flush()
        except Exception as e:
            print(f"Error publishing event: {e}")
            raise
